tls/s2n_early_data_io.c

```diff
@@ -117,3 +119,112 @@ S2N_RESULT s2n_early_data_validate_recv(struct s2n_connection *conn)
     RESULT_ENSURE(s2n_conn_get_current_message_type(conn) == END_OF_EARLY_DATA, S2N_ERR_EARLY_DATA_NOT_ALLOWED);
     return S2N_RESULT_OK;
 }
+
+static bool s2n_early_data_can_continue(struct s2n_connection *conn)
+{
+    uint32_t remaining_early_data_size = 0;
+    return s2n_connection_get_remaining_early_data_size(conn, &remaining_early_data_size) >= S2N_SUCCESS
+            && remaining_early_data_size > 0;
+}
+
+S2N_RESULT s2n_send_early_data_impl(struct s2n_connection *conn, const uint8_t *data, ssize_t data_len,
+        ssize_t *data_sent, s2n_blocked_status *blocked)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(blocked);
+    *blocked = S2N_NOT_BLOCKED;
+    RESULT_ENSURE_REF(data_sent);
+    *data_sent = 0;
+
+    RESULT_ENSURE(conn->mode == S2N_CLIENT, S2N_ERR_SERVER_MODE);
+
+    if (!s2n_early_data_can_continue(conn)) {
+        return S2N_RESULT_OK;
+    }
+
+    if (s2n_negotiate(conn, blocked) < S2N_SUCCESS) {
+        if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+            return S2N_RESULT_ERROR;
+        } else if (*blocked != S2N_BLOCKED_ON_EARLY_DATA && *blocked != S2N_BLOCKED_ON_READ) {
+            return S2N_RESULT_ERROR;
+        }
+    }
+
+    uint32_t early_data_to_send = 0;
+    RESULT_GUARD_POSIX(s2n_connection_get_remaining_early_data_size(conn, &early_data_to_send));
+    early_data_to_send = MIN(data_len, early_data_to_send);
+    if (early_data_to_send) {
+        ssize_t send_result = s2n_send(conn, data, early_data_to_send, blocked);
+        RESULT_GUARD_POSIX(send_result);
+        *data_sent = send_result;
+    }
+
+    if (s2n_negotiate(conn, blocked) < S2N_SUCCESS) {
+        if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+            return S2N_RESULT_ERROR;
+        } else if (*blocked == S2N_BLOCKED_ON_EARLY_DATA) {
+            return S2N_RESULT_OK;
+        } else if (s2n_early_data_can_continue(conn)) {
+            return S2N_RESULT_ERROR;
+        } else {
+            return S2N_RESULT_OK;
+        }
+    }
+    return S2N_RESULT_OK;
+}
+
+int s2n_send_early_data(struct s2n_connection *conn, const uint8_t *data, ssize_t data_len,
+        ssize_t *data_sent, s2n_blocked_status *blocked)
+{
+    POSIX_GUARD(s2n_connection_set_early_data_expected(conn));
+    s2n_result result = s2n_send_early_data_impl(conn, data, data_len, data_sent, blocked);
+    POSIX_GUARD(s2n_connection_set_end_of_early_data(conn));
```

Comment:
Eh IDK. If customers are calling it directly, `s2n_connection_set_end_of_early_data` makes more sense. It's just a little wonky here because we assume EndOfEarlyData unless told otherwise, so that we can support the proposed behavior of switching to s2n_negotiate to end early data.

Reply:
Does this mean a caller can only send a single chunk of early data? As in, they'll have to accumulate it all of it outside in a single, contiguous buffer before handing it to s2n?